### PR TITLE
[TT-6446] Add http transport pool, get enforced timeout from APISpec

### PR DESCRIPTION
**Pull request by titpetric**
https://tyktech.atlassian.net/browse/TT-6446

- Updates reverse proxy with go stdlib improvements
- Use httputil.BufferPool
- Implement carrying *APISpec on context
- Ensure http transport created timeout from APISpec

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit:  
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5998879602)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 4bb8599c3bd989b0fdbfb4af46d75bcf56598bf8 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5999318999)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 8ee893506919a24bd0fd602160949bf1523408da 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5999376846)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 2260f57e85faba2f41676fd1da6e7f9c96508f35 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5999577633)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 79afe9208dc41d10fa9fb101ed877d6ace7a769f 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5999669263)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 1d3af388d0189cd413d5b070ad5af13ab76d3a41 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6000280745)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: f1daa3d4f371b51909b2effb10c024efcd42f0fe 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6000633952)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 926d96d44f911378a21d01d60c86399f6a615f65 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6001556012)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: c1cf2907f9e38538977bfbf5a21f3cc6dd4c2218 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6004459760)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: e21e603c96b8dcff6db5231aba70d3ff0c16945e 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6004511118)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 0dafde16181518f06623cc2c4b4fa8d9621538fd 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6009256622)

File: gateway/reverse_proxy.go [hunk 1]
```diff
@@ -380,44 +384,56 @@ type ReverseProxy struct {
...
-	dialContextFunc := dialer.DialContext
-	if p.Gw.dnsCacheManager.IsCacheEnabled() {
-		dialContextFunc = p.Gw.dnsCacheManager.WrapDialer(dialer)
-	}
+	dialContextFunc := p.dialContextFunc(dialerTimeout)
+
+	gwConfig := p.Gw.GetConfig()
+
+	// Construct a key from values that apply to the transport.
+	key := fmt.Sprintf("%d;%d;%d", dialerTimeout, gwConfig.MaxIdleConns, gwConfig.MaxIdleConnsPerHost)
```
Comment (titpetric) [inline gateway/reverse_proxy.go:414]: As per discussion with @jeffy-mathew, using APISpec.ID makes more sense in here, to isolate transports better than just over a timeout value. Other fields inside transport may be configured differently (TLS, etc.).

File: gateway/reverse_proxy.go [hunk 2]
```diff
@@ -380,44 +384,56 @@ type ReverseProxy struct {
...
+	transportPool *httputil.TransportPool
+
 	logger *logrus.Entry
-	sp     sync.Pool
 	Gw     *Gateway `json:"-"`
 }
 
+// A BufferPool is an interface for getting and returning temporary
+// byte slices for use by io.CopyBuffer.
+type BufferPool interface {
```
Comment (titpetric) [inline gateway/reverse_proxy.go:396]: use httputil.BufferPool from the stdlib

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5474/merge
Commit: 420728d03e5f1034281a2b9c30648a1c2e021c83 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6048126626)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5474/merge
Commit: 04fd0739af135bd845b783c3025854363d738bd9 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6048244815)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5474/merge
Commit: 7d26669d7bf419e05eee8c61815ce9c2f44121e1 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6048422961)

File: internal/httputil/transport_pool.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	tp.mu.Lock()
+	defer tp.mu.Unlock()
+
+	prev, ok := tp.pool[key]
+
+	tp.pool[key] = transport
+
+	if ok {
+		defer func() {
+			go prev.CloseIdleConnections()
```
Comment (titpetric) [inline internal/httputil/transport_pool.go:34]: consolidate L60, invoke same code path

File: gateway/reverse_proxy.go (hunk 1 above)
Comment (titpetric) [reply gateway/reverse_proxy.go]: Using APIID now

File: internal/httputil/transport_pool.go (hunk 3 above)
Comment (titpetric) [reply internal/httputil/transport_pool.go:34]: @jeffy-mathew what should we do for spec.Release?

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5474/merge
Commit:  
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6083943205)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5474/merge
Commit: bb2d84ebfc90ba7e6dfbd9ecfad10c82a0f0a12c 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6084030026)

**Pull request closed without merging**